### PR TITLE
test(amf): stateless unit test

### LINKED ISSUE #11025: Stateless Feature unit test for AMF.

### PR DESCRIPTION
**Pull request by shashidhar-patil**
## Summary
Provides unit test for the stateless feature.

Approach: In the unit test, after the serialization process we do not write to the Redis DB, instead we write to an in-memory data store map defined in the AmfClientServicer. 

Implementation:
- Set the required config flags (use_stateless=true)
- In the test case, we trigger a TearDown() which basically clears the amf_app_desc data. Then we trigger a Setup() which internally restores back the values from the data store into the amf_app_desc.

**Note: 
Commit 1: Stateless Base code pulled from PR #10605
Commit 2: Stateless unit test.**

## Test Plan
- Verified basic sanity with ueransim.
- Verified with Unit test.
- Please find attached logs and pcap for basic registration

mme log:
[mme.log](https://github.com/magma/magma/files/7694585/mme.log)

pcap snap:
stateless_pcap

Unit test Snap:
stateless_ut

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by shashidhar-patil** [comment]
**PR rebased.**

mme log: [05/01/2022]
[mme.log](https://github.com/magma/magma/files/7812241/mme.log)

pcap snapshot: [05/01/2022]
stateless_ut_pcap_2

Unit Test: [05/01/2022]
stateless_ut_framework2

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:106]: Is the intent to build this during the unit testing and not during the normal build? Besides this, I also see that the flag is used inside `read_state_from_db()` already, why is it checked here again?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: should not this be `!MME_UNIT_TEST`?

Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: also again looks like you handle this build flag inside the call `read_state_from_db()`. it should not be here then.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:106]: Yes, the intent is to build this only during the unit testing and not during the normal build. Since this acts as a base for all the upcoming changes(`read_ue_state_from_db `does not have` #if !MME_UNIT_TEST `check at state manager class), this check acts as a double-check to make sure stateless changes do not affect the amf application code. This check is removed in the Stateless Integration PR( #11090 ) and only the check inside `read_state_from_db()` will stay.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Addressed this review comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Specify complete file path

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:106]: Could you please add a comment here that this is temporary and will be removed in future PR?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.h
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.h]:
```suggestion
constexpr char AMF_IMSI_UE_ID_TABLE_NAME[] = "amf_app_imsi_ue_id_htbl";
constexpr char AMF_TUN_UE_ID_TABLE_NAME[]  = "amf_app_tun11_ue_id_htbl";
constexpr char AMF_GUTI_UE_ID_TABLE_NAME[] = "amf_app_guti_ue_id_htbl";
constexpr char AMF_GNB_UE_ID_AMF_UE_ID_TABLE_NAME[] =
    "anf_app_gnb_ue_ngap_id_ue_id_htbl";
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Please use camel case for naming the test suite

```suggestion
TEST(TestAMFStateConverter, TestGutiToString) {
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
TEST(TestAMFStateConverter,TestStateToProto) {
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  EXPECT_FALSE(
      AMFClientServicer::getInstance().map_tableKey_protoStr.isEmpty());
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
EXPECT_TRUE(get_ue_id_from_imsi(amf_app_desc_p, imsi64, &ue_id));
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  EXPECT_EQ(rc, RETURNok);
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: It is better to use EXPECT_EQ instead of EXPECT_TRUE for comparing return code. Please change it every where in this test.

```suggestion
  EXPECT_EQ(rc, RETURNok);
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  EXPECT_FALSE(
      amf_app_desc_p->amf_ue_contexts.imsi_amf_ue_id_htbl.isEmpty());
  EXPECT_FALSE(
      amf_app_desc_p->amf_ue_contexts.tun11_ue_context_htbl.isEmpty());
  EXPECT_FALSE(
      amf_app_desc_p->amf_ue_contexts.guti_ue_context_htbl.isEmpty());
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:239]: Please add a comment that this section is simulating a service restart.

**Review by ssanadhya — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp:106]: Sure, added the required comment.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:239]: Added the required comment.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Review comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.h]: Review comment addressed.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Yes sure, will maintain this practice going forward. Review comment addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Review comment addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Addressed this review comment

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Review comment addressed.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Okay sure, made the required changes

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Review comment addressed

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Review comment addressed

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Please add the ToDo tag 

`// ToDo: This is a temporary check and will be removed in upcoming PR`

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:243]: Can we also include a check that amf_ngap_ue_id_ue_context_htbl is empty here?

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  /* Send ip address response from pipelined */
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  /* Send pdu session setup response from smf */
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]:
```suggestion
  /* Send pdu resource setup response from UE */
```

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:13]: Please review the lint warning

**Review by ssanadhya — COMMENTED** [review]
LGTM, with minor comments. 

Can be merged once revised and @ardzoht approves.

File: lte/gateway/c/core/oai/include/map.h
Comment (ardzoht) [inline lte/gateway/c/core/oai/include/map.h]:
```suggestion
    return umap.empty();
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: add `using magma::lte...` directive at top of the file and update this and other usages accordingly

File: lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h]: this should be defined on `map.h`

File: lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h]:
```suggestion
  map_string_string_t map_table_key_proto_str;
```

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:243]: In this, we are syncing only amf_app_desc whereas ue_context is not synced. The ue_context syncing and amf_ue_ngap_id_ue_context_htbl check is added as part of the Additional unit test cases PR( #11034).

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Review comment addressed.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Review comment addressed.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Review comment addressed.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp]: Review comment addressed.

File: lte/gateway/c/core/oai/include/map.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/include/map.h]: Review comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: Review comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h]: Review comment addressed.

File: lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h]: Review comment addressed.

File: lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp
Comment (shashidhar-patil) [reply lte/gateway/c/core/oai/test/amf/test_amf_stateless.cpp:13]: Resolved the warning.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp
Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_state_manager.cpp]: @shashidhar-patil , please address the Lint warning here.

**Review by ssanadhya — APPROVED** [review]
LGTM, can be merged after @ardzoht 's approval.

**Review by ardzoht — APPROVED** [review]
lgtm

**Pull request merged**
